src/System.Text.Primitives/System/Text/Encoding/Utf8/Utf8TextEncoding.cs

```diff
@@ -1,308 +1,96 @@
-
 namespace System.Text.Utf8
 {
     class Utf8TextEncoding : TextEncoder
     {
-        private const byte b0000_0111U = 0x07; //7
-        private const byte b0000_1111U = 0x0F; //15
-        private const byte b0001_1111U = 0x1F; //31
-        private const byte b0011_1111U = 0x3F; //63
-        private const byte b0111_1111U = 0x7F; //127
-        private const byte b1000_0000U = 0x80; //128
-        private const byte b1100_0000U = 0xC0; //192
-        private const byte b1110_0000U = 0xE0; //224
-        private const byte b1111_0000U = 0xF0; //240
-        private const byte b1111_1000U = 0xF8; //248
+        #region Decoding implementation
 
-        public override bool TryEncodeFromUtf8(ReadOnlySpan<byte> utf8, Span<byte> buffer, out int bytesWritten)
+        public override bool TryDecode(ReadOnlySpan<byte> data, out string text, out int bytesConsumed)
         {
-            if (buffer.Length < utf8.Length)
+            int len;
+            if (!Utf8Encoder.TryComputeStringLength(data, out len))
             {
-                bytesWritten = 0;
+                text = string.Empty;
+                bytesConsumed = 0;
                 return false;
             }
 
-            utf8.CopyTo(buffer);
-            bytesWritten = utf8.Length;
-            return true;
-        }
-
-        public override bool TryEncodeFromUtf16(ReadOnlySpan<char> utf16, Span<byte> buffer, out int bytesWritten)
-        {
-            var availableBytes = buffer.Length;
-            var inputLength = utf16.Length;
-            int tempCounter = 0;
-            bytesWritten = 0;
-
-            for (int i = 0; i < inputLength; i++)
-            {
-                var codePoint = utf16[i];
-
-                if (codePoint <= 0x7F)
-                {
-                    if (tempCounter + 1 > availableBytes)
-                    {
-                        bytesWritten = 0;
-                        return false;
-                    }
-                    buffer[tempCounter] = (byte)(codePoint);
-                    tempCounter++;
-                    continue;
-                }
-                else if (codePoint <= 0x7FF)
-                {
-                    if (tempCounter + 2 > availableBytes)
-                    {
-                        bytesWritten = 0;
-                        return false;
-                    }
-                    buffer[tempCounter] = (byte)(((codePoint >> 6) & b0001_1111U) | b1100_0000U);
-                    tempCounter++;
-                    buffer[tempCounter] = (byte)(((codePoint) & b0011_1111U) | b1000_0000U);
-                    tempCounter++;
-                    continue;
-                }
-                else if (char.IsSurrogate(codePoint))
-                {
-                    if (tempCounter + 4 > availableBytes)
-                    {
-                        bytesWritten = 0;
-                        return false;
-                    }
+            var buffer = new char[len];
```

Comment:
Since you know the string lengh, I would do the following:
```c#
var str = new string(' ', len);
fixed(char* p = string){
    var charSpan = new Span<char>(p, len);
}
```